sharktank/sharktank/tools/tuner/candidate_gen.py

```diff
@@ -0,0 +1,1386 @@
+        assert lhs_shaped_type.rank() == 4
+
+        rhs_shaped_type = parse_tensor_type(rhs_tensor_type)
+        assert rhs_shaped_type.rank() == 4
+
+        res_tensor_type = dps.group("RES")
+        res_shaped_type = parse_tensor_type(res_tensor_type)
+        assert res_shaped_type.rank() == 4
+
+        # int64_t n = outputShape[0];
+        # int64_t oh = outputShape[1];
+        # int64_t ow = outputShape[2];
+        # int64_t oc = outputShape[3];
+        # int64_t fh = filterShape[0];
+        # int64_t fw = filterShape[1];
+        # int64_t ic = filterShape[2];
+        dim_info = ConvDimInfo.from_rhs_res(rhs_shaped_type, res_shaped_type)
+        return ProblemSize(
+            MatmulSize(
+                M=dim_info.oh * dim_info.ow,
+                N=dim_info.oc,
+                K=dim_info.fh * dim_info.fw * dim_info.ic,
+                B=dim_info.n,
+            ),
+            lhs_shaped_type,
+            rhs_shaped_type,
+            res_shaped_type,
+            DispatchKind.conv,
+        )
+
+    assert False, "Shape not found"
+
+
+def get_shapes_contract(
+    template: list[str], lhs_dims: str, rhs_dims: str
+) -> ProblemSize:
+    for line in template:
+        if "linalg.generic" not in line:
+            continue
+        if "lowering_config =" not in line:
+            continue
+        if '"reduction"' not in line:
+            continue
+
+        # ins(%7, %8 : tensor<2x1024x1280xf16>, tensor<20x64x1280xf16>)
+        cont_re = rf"{MlirRegex.dps_ins_two_args()}\s+{MlirRegex.dps_outs_one_arg()}"
+        dps = re.search(cont_re, line)
+        if dps is None:
+            continue
+
+        lhs_tensor_type = dps.group("LHS")
+        rhs_tensor_type = dps.group("RHS")
+        lhs_shaped_type = parse_tensor_type(lhs_tensor_type)
+        assert lhs_shaped_type.rank() == len(lhs_dims)
+
+        rhs_shaped_type = parse_tensor_type(rhs_tensor_type)
+        assert rhs_shaped_type.rank() == len(rhs_dims)
+
+        res_tensor_type = dps.group("RES")
+        res_shaped_type = parse_tensor_type(res_tensor_type)
+        assert res_shaped_type.rank() >= 2
+
+        M = math.prod(
+            val if dim == "m" else 1
+            for dim, val in zip(lhs_dims, lhs_shaped_type.shape)
+        )
+        N = math.prod(
+            val if dim == "n" else 1
+            for dim, val in zip(rhs_dims, rhs_shaped_type.shape)
+        )
+        K0 = math.prod(
+            val if dim == "k" else 1
+            for dim, val in zip(lhs_dims, lhs_shaped_type.shape)
+        )
+        K1 = math.prod(
+            val if dim == "k" else 1
+            for dim, val in zip(rhs_dims, rhs_shaped_type.shape)
+        )
+        assert K0 == K1
+
+        return ProblemSize(
+            MatmulSize(M, N, K0),
+            lhs_type=lhs_shaped_type,
+            rhs_type=rhs_shaped_type,
+            res_type=res_shaped_type,
+            dispatch_kind=DispatchKind.contraction,
+        )
+
+    assert False, "Shape not found"
```

Comment:
Could include more information about what was being searched for in this message